File: app/houston-tauri/src/agent_store/activity.rs
//! CRUD operations for `.houston/activity.json`.

use super::helpers::{read_json, write_json};
use super::types::{Activity, ActivityUpdate};
use chrono::Utc;
use std::path::Path;
use uuid::Uuid;

const FILE: &str = "activity";

pub fn list(root: &Path) -> Result<Vec<Activity>, String> {
    read_json::<Vec<Activity>>(root, FILE)
}

pub fn create(
    root: &Path,
    title: &str,
    description: &str,
    agent: Option<&str>,
    worktree_path: Option<&str>,
) -> Result<Activity, String> {
    let mut items = list(root)?;
    let now = Utc::now().to_rfc3339();
    let item = Activity {
        id: Uuid::new_v4().to_string(),
        title: title.to_string(),
        description: description.to_string(),
        status: "running".to_string(),
        claude_session_id: None,
        session_key: None,
        agent: agent.map(|s| s.to_string()),
        worktree_path: worktree_path.map(|s| s.to_string()),
        routine_id: None,
        routine_run_id: None,
        updated_at: Some(now),
    };
    items.push(item.clone());
    write_json(root, FILE, &items)?;
    Ok(item)
}

pub fn update(root: &Path, id: &str, updates: ActivityUpdate) -> Result<Activity, String> {
    let mut items = list(root)?;
    let item = items
        .iter_mut()
        .find(|t| t.id == id)
        .ok_or_else(|| format!("Activity not found: {id}"))?;

    if let Some(title) = updates.title {
        item.title = title;
    }
    if let Some(description) = updates.description {
        item.description = description;
    }
    if let Some(status) = updates.status {
        item.status = status;
    }
    if let Some(session_id) = updates.claude_session_id {
        item.claude_session_id = session_id;
    }
    if let Some(session_key) = updates.session_key {
        item.session_key = Some(session_key);
    }
    if let Some(agent) = updates.agent {
        item.agent = Some(agent);
    }
    if let Some(worktree_path) = updates.worktree_path {
        item.worktree_path = worktree_path;
    }
    if let Some(routine_id) = updates.routine_id {
        item.routine_id = Some(routine_id);
    }
    if let Some(routine_run_id) = updates.routine_run_id {
        item.routine_run_id = Some(routine_run_id);
    }

    item.updated_at = Some(Utc::now().to_rfc3339());

    let result = item.clone();
    write_json(root, FILE, &items)?;
    Ok(result)
}

pub fn delete(root: &Path, id: &str) -> Result<(), String> {
    let mut items = list(root)?;
    let before = items.len();
    items.retain(|t| t.id != id);
    if items.len() == before {
        return Err(format!("Activity not found: {id}"));
    }
    write_json(root, FILE, &items)
}

File: app/src/data/activity.ts
/**
 * `.houston/activity/activity.json` — the board.
 *
 * Schema-validated via `@houston-ai/agent-schemas/activity.schema.json`.
 * Written atomically on every mutation (the backend handles the temp-file + rename).
 */

import schema from "@houston-ai/agent-schemas/activity.schema.json";
import { newId, now, readAgentJson, writeAgentJson } from "./agent-file";

/**
 * Mirrors the engine's `ActivityStatus` enum (engine/houston-engine-core
 * /src/agents/status.rs). Authoritative source is the Rust side; this
 * union exists so the typechecker catches drift at compile time.
 */
export type ActivityStatus =
  | "queued"
  | "running"
  | "needs_you"
  | "done"
  | "error"
  | "cancelled"
  | "interrupted";

/**
 * Durability ownership token. Present only on in-flight (queued/running)
 * rows; the engine clears it on every terminal transition. Mirrors
 * `engine/houston-engine-core/src/agents/lease.rs`.
 */
export interface ActivityLease {
  lease_id: string;
  owner_pid: number;
  expires_at: string;
}

export interface Activity {
  id: string;
  title: string;
  description: string;
  status: ActivityStatus;
  lease?: ActivityLease | null;
  claude_session_id?: string | null;
  session_key?: string;
  agent?: string;
  worktree_path?: string | null;
  routine_id?: string;
  routine_run_id?: string;
  updated_at?: string;
  provider?: string;
  model?: string;
}

export interface ActivityUpdate {
  title?: string;
  description?: string;
  status?: ActivityStatus;
  claude_session_id?: string | null;
  session_key?: string;
  agent?: string;
  worktree_path?: string | null;
  routine_id?: string;
  routine_run_id?: string;
  provider?: string;
  model?: string;
}

const NAME = "activity";
const s = schema as unknown as Parameters<typeof readAgentJson>[2];

export async function list(agentPath: string): Promise<Activity[]> {
  return readAgentJson<Activity[]>(agentPath, NAME, s, []);
}

export async function create(
  agentPath: string,
  title: string,
  description = "",
  agent?: string,
  worktreePath?: string,
  provider?: string,
  model?: string,
): Promise<Activity> {
  const items = await list(agentPath);
  const item: Activity = {
    id: newId(),
    title,
    description,
    // Born `queued` so the engine's reaper never sees a `running` row
    // without an attached lease — `sessions::start` is responsible for
    // promoting to `running` and minting the lease at the same time.
    status: "queued",
    claude_session_id: null,
    agent,
    worktree_path: worktreePath ?? null,
    updated_at: now(),
    provider,
    model,
  };
  await writeAgentJson(agentPath, NAME, s, [...items, item]);
  return item;
}

export async function update(
  agentPath: string,
  id: string,
  patch: ActivityUpdate,
): Promise<Activity> {
  const items = await list(agentPath);
  const idx = items.findIndex((a) => a.id === id);
  if (idx === -1) throw new Error(`Activity not found: ${id}`);
  const merged: Activity = {
    ...items[idx],
    ...patch,
    updated_at: now(),
  };
  const next = [...items];
  next[idx] = merged;
  await writeAgentJson(agentPath, NAME, s, next);
  return merged;
}

export async function remove(agentPath: string, id: string): Promise<void> {
  const items = await list(agentPath);
  const next = items.filter((a) => a.id !== id);
  if (next.length === items.length) throw new Error(`Activity not found: ${id}`);
  await writeAgentJson(agentPath, NAME, s, next);
}
